src/historywork/DownloadApplyTxsWork.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/DownloadApplyTxsWork.h"
+#include "catchup/ApplyCheckpointWork.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryManager.h"
+#include "historywork/GetAndUnzipRemoteFileWork.h"
+#include "ledger/LedgerManager.h"
+#include "work/WorkSequence.h"
+
+namespace stellar
+{
+
+DownloadApplyTxsWork::DownloadApplyTxsWork(
+    Application& app, TmpDir const& downloadDir, LedgerRange range,
+    LedgerHeaderHistoryEntry& lastApplied)
+    : Work(app, "download-apply-ledgers", RETRY_NEVER)
+    , mRange(range)
+    , mDownloadDir(downloadDir)
+    , mLastApplied(lastApplied)
+    , mCurrCheckpoint(
+          app.getHistoryManager().checkpointContainingLedger(range.first()))
+{
+}
+
+DownloadApplyTxsWork::~DownloadApplyTxsWork()
+{
+}
+
+BasicWork::State
+DownloadApplyTxsWork::doWork()
+{
+    // This will properly abort the remaining children
+    if (anyChildRaiseFailure())
+    {
+        for (auto const& w : mRunning)
+        {
+            w->wakeUp();
+        }
+        return State::WORK_FAILURE;
+    }
+
+    auto prev = mRunning.begin();
+    for (auto childIt = mRunning.begin(); childIt != mRunning.end();)
```

Comment:
I think this is more complicated than it should be.

I am pretty sure that you can use `BatchWork` to perform what you're trying to do:
you can yield `download-apply` sequences (gated by the the last yielded value if there was one)

Reply:
I think what you are suggesting is to have children asynchronously wait, while my implementation makes parent responsible for waking up an appropriate child. I thought about the approach you suggested when first implementing it (in fact, initially DownloadApplyTxsWork was inheriting from BatchWork). But this approach seemed less efficient and safe. Since we're doing it in batches, there will be a bunch of works waiting for each other, and potentially waking up and going back to sleep many times before the condition is satisfied (for example, applying 33 checkpoints with Postgres takes ~1.5 min on my machine, meaning last work in the batch will need to wake up almost 100 times). On the other hand, if some checkpoints apply really fast, dependent work still needs to wait for a second, whereas it could have proceeded earlier.
As to safety, parent work should keep references to whichever works are running and clean it up when needed (note that the condition to when it is okay to delete a child differs from ones of Work and BatchWork - those just delete a child once it succeeds or fails. In our case, we can only delete a child when it succeeds _and_ its dependent work succeeds (as the condition function only captures a weak pointer, so work must be kept alive).
The considerations above made me think that having parent centrally take care of its gated works is simpler and safer. Let me know what you think!

Reply:
It looks more and more that true DAG will be needed at some time in a future. What do you think about implementing it now?

Reply:
I thought about it as well: as mentioned in the PR description, I did not find a good reason to introduce DAG at this time: we only have 1 use case for this, and I anticipate the new validation logic will have the same use case.

Reply:
I was not suggesting changing dependencies, just reusing code:
right now `DownloadApplyTxsWork` is duplicating the logic in `BatchWork` for no particular good reason (throttling, cleaning up of "completed" work). I don't see why using `BatchWork` would cause delays/sleeps.

Reason maybe you could not make it work was because of the conditions?
I think you can use regular `shared_ptr` to the previous work instead of weak pointers as we're not introducing cycles. This allows to let `BatchWork` manage the list of work as it wants and leave your own logic do what it wants as well.